File: devices-api-java-mvc/src/test/java/com/lucas/devicesapijavamvc/api/DeviceControllerTest.java
package com.lucas.devicesapijavamvc.api;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.lucas.devicesapijavamvc.api.dto.DeviceRequest;
import com.lucas.devicesapijavamvc.domain.Device;
import com.lucas.devicesapijavamvc.domain.DeviceState;
import com.lucas.devicesapijavamvc.service.DeviceService;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.test.util.ReflectionTestUtils;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.context.bean.override.mockito.MockitoBean;

import java.time.OffsetDateTime;
import java.util.List;
import java.util.UUID;

import static org.hamcrest.Matchers.hasSize;
import static org.hamcrest.Matchers.is;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(controllers = DeviceController.class)
class DeviceControllerTest {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ObjectMapper objectMapper;

    @MockitoBean
    private DeviceService deviceService;

    private Device device(Long id,
                          UUID externalId,
                          String name,
                          String brand,
                          DeviceState state,
                          OffsetDateTime creationTime) {

        Device d = Device.fromEvent(
                externalId,
                name,
                brand,
                state,
                creationTime
        );
        ReflectionTestUtils.setField(d, "id", id);
        return d;
    }

    private Device device(Long id, String name, String brand, DeviceState state) {
        UUID externalId = UUID.randomUUID();
        OffsetDateTime creationTime = OffsetDateTime.now();
        return device(id, externalId, name, brand, state, creationTime);
    }

    @Test
    @DisplayName("POST /api/devices - creates device and returns 201 with body")
    void createDevice() throws Exception {
        DeviceRequest request = new DeviceRequest("Phone", "Apple", DeviceState.AVAILABLE);

        Device saved = device(
                1L,
                UUID.randomUUID(),
                request.name(),
                request.brand(),
                request.state(),
                OffsetDateTime.now()
        );

        when(deviceService.create(any(Device.class)))
                .thenReturn(saved);

        mockMvc.perform(post("/api/devices")
                        .contentType("application/json")
                        .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isCreated())
                .andExpect(jsonPath("$.id", is(saved.getId().intValue())))
                .andExpect(jsonPath("$.name", is(saved.getName())))
                .andExpect(jsonPath("$.brand", is(saved.getBrand())))
                .andExpect(jsonPath("$.state", is(saved.getState().name())));

        verify(deviceService).create(any(Device.class));
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("GET /api/devices/{id} - returns device when found")
    void getDevice() throws Exception {
        Device device = device(1L, "Router", "Cisco", DeviceState.IN_USE);

        when(deviceService.get(1L)).thenReturn(device);

        mockMvc.perform(get("/api/devices/1"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id", is(1)))
                .andExpect(jsonPath("$.name", is("Router")))
                .andExpect(jsonPath("$.brand", is("Cisco")))
                .andExpect(jsonPath("$.state", is("IN_USE")));

        verify(deviceService).get(1L);
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("GET /api/devices/{id} - returns 404 when not found")
    void getDeviceNotFound() throws Exception {
        when(deviceService.get(999L))
                .thenThrow(new DeviceService.NotFoundException("Device 999 not found"));

        mockMvc.perform(get("/api/devices/999"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.error", is("NOT_FOUND")))
                .andExpect(jsonPath("$.message", is("Device 999 not found")));

        verify(deviceService).get(999L);
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("GET /api/devices - returns all devices when no filters")
    void listAllDevices() throws Exception {
        List<Device> devices = List.of(
                device(1L, "Phone", "Apple", DeviceState.AVAILABLE),
                device(2L, "Tablet", "Samsung", DeviceState.RETIRED)
        );

        when(deviceService.all()).thenReturn(devices);

        mockMvc.perform(get("/api/devices"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", hasSize(2)))
                .andExpect(jsonPath("$[0].name", is("Phone")))
                .andExpect(jsonPath("$[1].name", is("Tablet")));

        verify(deviceService).all();
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("GET /api/devices?brand= - filters by brand")
    void listByBrand() throws Exception {
        List<Device> devices = List.of(
                device(1L, "Phone", "Apple", DeviceState.AVAILABLE)
        );

        when(deviceService.byBrand("Apple")).thenReturn(devices);

        mockMvc.perform(get("/api/devices")
                        .param("brand", "Apple"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", hasSize(1)))
                .andExpect(jsonPath("$[0].brand", is("Apple")));

        verify(deviceService).byBrand("Apple");
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("GET /api/devices?state= - filters by state")
    void listByState() throws Exception {
        List<Device> devices = List.of(
                device(1L, "Phone", "Apple", DeviceState.AVAILABLE)
        );

        when(deviceService.byState(DeviceState.AVAILABLE)).thenReturn(devices);

        mockMvc.perform(get("/api/devices")
                        .param("state", "AVAILABLE"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$", hasSize(1)))
                .andExpect(jsonPath("$[0].state", is("AVAILABLE")));

        verify(deviceService).byState(DeviceState.AVAILABLE);
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("PUT /api/devices/{id} - full update")
    void fullUpdate() throws Exception {
        DeviceRequest request = new DeviceRequest("New Phone", "Apple", DeviceState.IN_USE);

        Device updated = device(
                1L,
                UUID.randomUUID(),
                request.name(),
                request.brand(),
                request.state(),
                OffsetDateTime.now()
        );

        when(deviceService.updateFull(eq(1L), any(Device.class)))
                .thenReturn(updated);

        mockMvc.perform(put("/api/devices/1")
                        .contentType("application/json")
                        .content(objectMapper.writeValueAsString(request)))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id", is(1)))
                .andExpect(jsonPath("$.name", is("New Phone")))
                .andExpect(jsonPath("$.state", is("IN_USE")));

        verify(deviceService).updateFull(eq(1L), any(Device.class));
        verifyNoMoreInteractions(deviceService);
    }

    @Test
    @DisplayName("DELETE /api/devices/{id} - deletes device")
    void deleteDevice() throws Exception {
        doNothing().when(deviceService).delete(1L);

        mockMvc.perform(delete("/api/devices/1"))
                .andExpect(status().isNoContent());

        verify(deviceService).delete(1L);
        verifyNoMoreInteractions(deviceService);
    }
}

File: devices-api-kotlin-reactive/src/test/kotlin/com/lucas/devicesapikotlinreactive/infrastructure/api/DeviceControllerTest.kt
package com.lucas.devicesapikotlinreactive.infrastructure.api

import com.lucas.devicesapikotlinreactive.application.service.DeviceService
import com.lucas.devicesapikotlinreactive.domain.model.Device
import com.lucas.devicesapikotlinreactive.domain.model.DeviceState
import com.lucas.devicesapikotlinreactive.infrastructure.api.dto.DeviceRequest
import com.lucas.devicesapikotlinreactive.infrastructure.api.dto.DeviceResponse
import io.mockk.clearAllMocks
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test
import org.springframework.http.MediaType
import org.springframework.test.web.reactive.server.WebTestClient
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

class DeviceControllerTest {

    private val service = mockk<DeviceService>()
    private val controller = DeviceController(service)
    private val client: WebTestClient = WebTestClient.bindToController(controller)
        .controllerAdvice(RestExceptionHandler())
        .build()

    @AfterEach
    fun tearDown() = clearAllMocks()

    @Test
    @DisplayName("POST /api/devices creates a device and returns 201 with body")
    fun createShouldReturnCreatedDevice() {
        val request = DeviceRequest(
            name = "iPhone 15",
            brand = "Apple",
            state = DeviceState.AVAILABLE
        )

        val device = Device.create(
            name = request.name,
            brand = request.brand,
            state = request.state
        ).copy(id = 1L)

        every { service.create(request.name, request.brand, request.state) } returns Mono.just(device)

        val result = client.post()
            .uri("/api/devices")
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(request)
            .exchange()
            .expectStatus().isCreated
            .expectBody(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result.id).isEqualTo(1L)
        assertThat(result.name).isEqualTo("iPhone 15")
        assertThat(result.brand).isEqualTo("Apple")
        assertThat(result.state).isEqualTo(DeviceState.AVAILABLE.name)

        verify(exactly = 1) {
            service.create(request.name, request.brand, request.state)
        }
    }

    @Test
    @DisplayName("GET /api/devices/{id} returns device when found")
    fun getShouldReturnDeviceWhenFound() {
        val id = 1L
        val device = Device.create("Galaxy S24", "Samsung", DeviceState.IN_USE)
            .copy(id = id)

        every { service.get(id) } returns Mono.just(device)

        val result = client.get()
            .uri("/api/devices/{id}", id)
            .exchange()
            .expectStatus().isOk
            .expectBody(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result.id).isEqualTo(id)
        assertThat(result.name).isEqualTo("Galaxy S24")
        assertThat(result.brand).isEqualTo("Samsung")
        assertThat(result.state).isEqualTo(DeviceState.IN_USE.name)

        verify(exactly = 1) { service.get(id) }
    }

    @Test
    @DisplayName("GET /api/devices/{id} returns 404 when service throws NotFoundException")
    fun getShouldReturn404WhenNotFound() {
        val id = 42L
        every { service.get(id) } returns Mono.error(
            DeviceService.NotFoundException("Device $id not found")
        )

        client.get()
            .uri("/api/devices/{id}", id)
            .exchange()
            .expectStatus().isNotFound
            .expectBody()
            .jsonPath("$.error").isEqualTo("NOT_FOUND")
            .jsonPath("$.message").isEqualTo("Device 42 not found")

        verify(exactly = 1) { service.get(id) }
    }

    @Test
    @DisplayName("GET /api/devices without filters returns all devices")
    fun listShouldReturnAllDevicesWhenNoFilters() {
        val devices = listOf(
            Device.create("Pixel 9", "Google", DeviceState.AVAILABLE).copy(id = 1L),
            Device.create("Moto Edge", "Motorola", DeviceState.RETIRED).copy(id = 2L)
        )

        every { service.all() } returns Flux.fromIterable(devices)

        val result = client.get()
            .uri("/api/devices")
            .exchange()
            .expectStatus().isOk
            .expectBodyList(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result).hasSize(2)
        assertThat(result[0].name).isEqualTo("Pixel 9")
        assertThat(result[1].name).isEqualTo("Moto Edge")

        verify(exactly = 1) { service.all() }
    }

    @Test
    @DisplayName("GET /api/devices?brand= filters by brand")
    fun listShouldFilterByBrand() {
        val brand = "Apple"
        val devices = listOf(
            Device.create("iPhone 14", brand, DeviceState.AVAILABLE).copy(id = 1L),
            Device.create("iPhone 15", brand, DeviceState.IN_USE).copy(id = 2L)
        )

        every { service.byBrand(brand) } returns Flux.fromIterable(devices)

        val result = client.get()
            .uri { uriBuilder ->
                uriBuilder.path("/api/devices")
                    .queryParam("brand", brand)
                    .build()
            }
            .exchange()
            .expectStatus().isOk
            .expectBodyList(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result).hasSize(2)
        assertThat(result[0].brand).isEqualTo(brand)
        assertThat(result[1].brand).isEqualTo(brand)

        verify(exactly = 1) { service.byBrand(brand) }
    }

    @Test
    @DisplayName("GET /api/devices?state= filters by state")
    fun listShouldFilterByState() {
        val state = DeviceState.IN_USE
        val devices = listOf(
            Device.create("iPhone 13", "Apple", state).copy(id = 1L),
            Device.create("Galaxy S23", "Samsung", state).copy(id = 2L)
        )

        every { service.byState(state) } returns Flux.fromIterable(devices)

        val result = client.get()
            .uri { uriBuilder ->
                uriBuilder.path("/api/devices")
                    .queryParam("state", state)
                    .build()
            }
            .exchange()
            .expectStatus().isOk
            .expectBodyList(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result).hasSize(2)
        assertThat(result[0].state).isEqualTo(state.name)
        assertThat(result[1].state).isEqualTo(state.name)

        verify(exactly = 1) { service.byState(state) }
    }

    @Test
    @DisplayName("PUT /api/devices/{id} performs a full update and returns updated device")
    fun fullUpdateShouldReturnUpdatedDevice() {
        val id = 10L
        val request = DeviceRequest(
            name = "Updated Name",
            brand = "Updated Brand",
            state = DeviceState.AVAILABLE
        )

        val updatedDevice = Device.create(
            name = request.name,
            brand = request.brand,
            state = request.state
        ).copy(id = id)

        every { service.updateFull(id, any()) } returns Mono.just(updatedDevice)

        val result = client.put()
            .uri("/api/devices/{id}", id)
            .contentType(MediaType.APPLICATION_JSON)
            .bodyValue(request)
            .exchange()
            .expectStatus().isOk
            .expectBody(DeviceResponse::class.java)
            .returnResult()
            .responseBody!!

        assertThat(result.id).isEqualTo(id)
        assertThat(result.name).isEqualTo("Updated Name")
        assertThat(result.brand).isEqualTo("Updated Brand")
        assertThat(result.state).isEqualTo(DeviceState.AVAILABLE.name)

        verify(exactly = 1) { service.updateFull(id, any()) }
    }

    @Test
    @DisplayName("DELETE /api/devices/{id} returns 204 No Content")
    fun deleteShouldReturnNoContent() {
        val id = 99L
        every { service.delete(id) } returns Mono.empty()

        client.delete()
            .uri("/api/devices/{id}", id)
            .exchange()
            .expectStatus().isNoContent

        verify(exactly = 1) { service.delete(id) }
    }
}
